fpmsyncd/bgpmon.py

```diff
@@ -0,0 +1,171 @@
+        self.db.connect(self.db.STATE_DB, False)
+        self.db.delete_all_by_pattern(self.db.STATE_DB, "NEIGH_STATE_TABLE|*" )
+
+    # A quick way to check if there are anything happening within BGP is to
+    # check its log file has any activities. This is by checking its modified
+    # timestamp against the cached timestamp that we keep and if there is a
+    # difference, there is activity detected. In case the log file got wiped
+    # out, it will default back to constant pulling every 15 seconds
+    def bgp_activity_detected(self):
+        try:
+            timestamp = os.stat("/var/log/frr/frr.log").st_mtime
+            if timestamp != self.cached_timestamp:
+                #syslog.syslog("bgpmon timestamp diff detected: {}".format(timestamp))
+                self.cached_timestamp = timestamp
+                return True
+            else:
+                return False
+        except Exception, e:
+            return True
+
+    # Get a new snapshot of BGP neighbors and store them in the "new" location
+    def get_all_neigh_states(self):
+        # Clean up the "new" dictionaries
+        del self.new_ipv4_n[:]
+        self.new_ipv4_n_state.clear()
+        del self.new_ipv6_n[:]
+        self.new_ipv6_n_state.clear()
+        try:
+            cmd = "vtysh -c 'show bgp summary json'"
+            output = commands.getoutput(cmd)
+            peer_info = json.loads(output)
+            if "ipv4Unicast" in peer_info and "peers" in peer_info["ipv4Unicast"]:
+                self.new_ipv4_n = peer_info["ipv4Unicast"]["peers"].keys()
+                for i in range (0, len(self.new_ipv4_n)):
+                    self.new_ipv4_n_state[self.new_ipv4_n[i]] = \
+                    peer_info["ipv4Unicast"]["peers"][self.new_ipv4_n[i]]["state"]
+
+            if "ipv6Unicast" in peer_info and "peers" in peer_info["ipv6Unicast"]:
+                self.new_ipv6_n = peer_info["ipv6Unicast"]["peers"].keys()
+                for i in range (0, len(self.new_ipv6_n)):
+                    self.new_ipv6_n_state[self.new_ipv6_n[i]] = \
+                    peer_info["ipv6Unicast"]["peers"][self.new_ipv6_n[i]]["state"]
+
+        except Exception:
+            self.retrieve_fail_cnt += 1
+            syslog.syslog(syslog.LOG_ERR, "*ERROR* get_all_neigh_states Exception: %s"
+                    % (traceback.format_exc()))
+
+    def update_neigh_states(self):
+        # handle IPV4 case
+        for i in range (0, len(self.new_ipv4_n)):
+            neighb = self.new_ipv4_n[i]
+            key = "NEIGH_STATE_TABLE|%s" % neighb
+            if neighb in self.ipv4_n:
+                # only update the entry if sate changed
+                if self.ipv4_n_state[neighb] != self.new_ipv4_n_state[neighb]:
+                    # state changed. Update state DB for this entry
+                    state = self.new_ipv4_n_state[neighb]
+                    self.db.set(self.db.STATE_DB, key, 'state', state)
```

Comment:
>self.db.set [](start = 20, length = 11)

It will be more efficient to group all the DB operations in `update_neigh_states()` into pipeline.

Reply:
@qiluo-msft  Not sure if I understood your suggestion. Are you referring as to why not just overlay all the peer states in the stateDB without the need to check if there are any changes?  The reason is that in most of the time maybe a few may have changed/added/deleted in the new snapshot. If there are a lot of BGP peers in the DB, the unnecessary write that has no state changes can unnecessary busy the StateDB. let along the snmp proxy agent is trying to read this same table so if bgpmon locks it up for a longer time than necessary, it will have performance consequences for snmp proxy agent or any future agents that like to read these BGP peer stateDB entries. This is why it avoids unnecessary updates if the existing peer has no changes between the new snapshot and the previous cached state.  The desired design is to ensure that we favor efficiency for the "normal" steady state scenarios.  Please clarify If I misunderstood your suggestion.
Please clarify if I understood your

Reply:
No. I am talking about [https://github.com/andymccurdy/redis-py#pipelines](https://github.com/andymccurdy/redis-py#pipelines). And we have similar feature in swsscommon library.

---
In reply to: [484677456](https://github.com/Azure/sonic-swss/pull/1429#discussion_r484677456) [](ancestors = 484677456)

Reply:
@qiluo-msft Yes I agree.  thanks for the reference.  with pipelines we can batch several StateDB operations into one single transaction and thus cuts down some overhead. I think there is one PR (https://github.com/Azure/sonic-py-swsssdk/pull/85) that started to use this pipeline model.  I would prefer making this performance enhancement with a separate PR.  One reason is to observe if there are any potential pitfall from using pipelines until we gain some more usage experiences.  The other reason is that we first make sure that the basic design works first and then take on more enhancement improvements.

Reply:
@qiluo-msft I have changed the code to use redis pipeline for the StateDB operations.